src/Physics/Evaluate_friction_law.f90

```diff
@@ -1457,4 +1471,106 @@ SUBROUTINE rate_and_state_nuc103(TractionGP_XY,TractionGP_XZ,        & ! OUT: tr
   !
  END SUBROUTINE rate_and_state_nuc103
 
+   SUBROUTINE updateStateVariable (nBndGP, RS_f0, RS_b, RS_a, RS_sr0, RS_fw, RS_srW, RS_sl0, &
+                         SV0, time_inc, SR_tmp, LocSV)
+    !-------------------------------------------------------------------------!
+    IMPLICIT NONE
+    !-------------------------------------------------------------------------!
+    ! Argument list declaration
+    INTEGER                  :: nBndGP
+    REAL                     :: RS_f0, RS_b, RS_a(nBndGP), RS_sr0, RS_fw, RS_srW(nBndGP), RS_sl0(nBndGP) !constant input parameters
+    REAL                     :: SV0(nBndGP), time_inc, SR_tmp(nBndGP)                  !changing during iterations
+    REAL                     :: flv(nBndGP), fss(nBndGP), SVss(nBndGP), LocSV(nBndGP)                 !calculated in this routine
+    !-------------------------------------------------------------------------!
+    INTENT(IN)    :: RS_f0, RS_b, RS_a, RS_sr0, RS_fw, RS_srW, &
+                     RS_sl0, SV0, time_inc, SR_tmp
+    INTENT(INOUT) :: LocSV
+    !-------------------------------------------------------------------------!
+
+    ! low-velocity steady state friction coefficient
+    flv = RS_f0 - (RS_b-RS_a)* LOG(SR_tmp/RS_sr0)
+    ! steady state friction coefficient
+    fss = RS_fw + (flv - RS_fw)/(1.0D0+(SR_tmp/RS_srW)**8d0)**(1.0D0/8.0D0)
+    ! steady-state state variable
+    ! For compiling reasons we write SINH(X)=(EXP(X)-EXP(-X))/2
+    SVss = RS_a * LOG(2.0D0*RS_sr0/SR_tmp * (EXP(fss/RS_a)-EXP(-fss/RS_a))/2.0D0)
+
+    ! exact integration of dSV/dt DGL, assuming constant V over integration step
+    LocSV = Svss*(1.0D0-EXP(-SR_tmp*time_inc/RS_sl0))+EXP(-SR_tmp*time_inc/RS_sl0)*SV0
+
+
+    IF (ANY(IsNaN(LocSV)) .EQV. .TRUE.) THEN
+       logError(*) 'NaN detected'
+       STOP
+    ENDIF
+
+
+  END SUBROUTINE updateStateVariable
+
+  SUBROUTINE IterativelyInvertSR (nBndGP, nSRupdates, LocSR, RS_sr0, LocSV, RS_a, &
+                             n_stress, sh_stress, invZ, SRtest, has_converged)
+    !-------------------------------------------------------------------------!
+    IMPLICIT NONE
+    !-------------------------------------------------------------------------!
+    ! Argument list declaration
+    LOGICAL       :: has_converged                                            !check convergence
+    INTEGER       :: nSRupdates, i, nBndGP
+    REAL          :: RS_sr0, RS_a(nBndGP)                                     !constants
+    REAL          :: SRtest(nBndGP), LocSR(nBndGP), LocSV(nBndGP)
+    REAL          :: n_stress(nBndGP), sh_stress(nBndGP), invZ
+    REAL          :: NR(nBndGP), dNR(nBndGP), tmp(nBndGP), tmp2(nBndGP), tmp3(nBndGP)
+    REAL          :: mu_f(nBndGP), dmu_f(nBndGP)                              !calculated here in routine
+    REAL          :: AlmostZero = 1D-8, aTolF = 5e-14
```

Comment:
same here?

Reply:
I think this does not make sense and will lead to a slower code, because if the converged value of SR is let say 10e-20, and we reinitialize every time SR at 1e-8, then we loop over the 60 iterations. That's might be why the TP code is about 30% slower (I just tested on something close to tpv103, order3).
I think aTolF should be one or 2 order of magnitude larger than AlmostZero to avoid that.

Reply:
What about AlmostZero = 1D-12, aTolF = 1e-8
or AlmostZero = 5D-18, aTolF = 5e-14?
Can you test that and see if it works?